File: frontend/src/api/interviews.ts
import { apiClient } from './client';
import type { Interview } from '../types/interview';

export type InterviewCreatePayload = {
  application_id: number;
  job_id: number;
  candidate_id: number;
  round?: number;
  interview_type?: string;
  scheduled_start_at?: string;
  scheduled_end_at?: string;
  location?: string;
  meeting_link?: string;
  interviewer_ids?: number[];
};

export async function fetchInterviews() {
  const { data } = await apiClient.get<Interview[]>('/interviews');
  return data;
}

export async function createInterview(payload: InterviewCreatePayload) {
  const { data } = await apiClient.post<Interview>('/interviews', payload);
  return data;
}

File: backend/app/api/routes/interviews.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.db.session import get_db
from app.models.application import JobApplication
from app.models.candidate import Candidate
from app.models.interview import Interview, InterviewInterviewer
from app.models.job import Job
from app.schemas.interview import InterviewCreate, InterviewRead

router = APIRouter()


@router.get("", response_model=list[InterviewRead])
def list_interviews(db: Session = Depends(get_db)) -> list[Interview]:
    return list(db.scalars(select(Interview).order_by(Interview.created_at.desc())).all())


@router.post("", response_model=InterviewRead)
def create_interview(payload: InterviewCreate, db: Session = Depends(get_db)) -> Interview:
    application = db.get(JobApplication, payload.application_id)
    if not application:
        raise HTTPException(status_code=404, detail="Application not found")
    if not db.get(Job, payload.job_id):
        raise HTTPException(status_code=404, detail="Job not found")
    if not db.get(Candidate, payload.candidate_id):
        raise HTTPException(status_code=404, detail="Candidate not found")
    if application.job_id != payload.job_id or application.candidate_id != payload.candidate_id:
        raise HTTPException(status_code=400, detail="Interview does not match application")
    data = payload.model_dump(exclude={"interviewer_ids"})
    interview = Interview(**data)
    db.add(interview)
    db.flush()
    for user_id in payload.interviewer_ids:
        db.add(InterviewInterviewer(interview_id=interview.id, user_id=user_id))
    db.commit()
    db.refresh(interview)
    return interview
